File: packages/Recognizers/ConversationLanguageUnderstanding/dotnet/CluApplication.cs
using System;

namespace Microsoft.Bot.Components.Recognizers.CLURecognizer
{
    /// <summary>
    /// Data describing a CLU application.
    /// </summary>
    public class CluApplication
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="CluApplication"/> class.
        /// </summary>
        /// <param name="projectName">CLU project name.</param>
        /// <param name="endpointKey">CLU subscription or endpoint key.</param>
        /// <param name="endpoint">CLU endpoint to use.</param>
        /// <param name="deploymentName">CLU deployment name.</param>
        public CluApplication(string projectName, string endpointKey, string endpoint, string deploymentName)
            : this((projectName, endpointKey, endpoint, deploymentName))
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="CluApplication"/> class.
        /// For unit tests only.
        /// </summary>
        protected CluApplication()
        {
        }

        private CluApplication(ValueTuple<string, string, string, string> props)
        {
            var (projectName, endpointKey, endpoint, deploymentName) = props;

            if (string.IsNullOrWhiteSpace(projectName))
            {
                throw new ArgumentException($"CLU \"{nameof(projectName)}\" parameter cannot be null or empty.");
            }

            if (!Guid.TryParse(endpointKey, out var _))
            {
                throw new ArgumentException($"\"{endpointKey}\" is not a valid CLU subscription key.");
            }

            if (string.IsNullOrWhiteSpace(endpoint))
            {
                throw new ArgumentException($"CLU \"{nameof(endpoint)}\" parameter cannot be null or empty.");
            }

            if (!Uri.IsWellFormedUriString(endpoint, UriKind.Absolute))
            {
                throw new ArgumentException($"\"{endpoint}\" is not a valid CLU endpoint.");
            }

            if (string.IsNullOrWhiteSpace(deploymentName))
            {
                throw new ArgumentException($"CLU \"{nameof(deploymentName)}\" parameter cannot be null or empty.");
            }

            ProjectName = projectName;
            EndpointKey = endpointKey;
            Endpoint = endpoint;
            DeploymentName = deploymentName;
        }

        /// <summary>
        /// Gets or sets CLU project name.
        /// </summary>
        /// <value>
        /// CLU project name.
        /// </value>
        public string ProjectName { get; set; } = default!;

        /// <summary>
        /// Gets or sets CLU subscription or endpoint key.
        /// </summary>
        /// <value>
        /// CLU subscription or endpoint key.
        /// </value>
        public string EndpointKey { get; set; } = default!;

        /// <summary>
        /// Gets or sets CLU endpoint.
        /// </summary>
        /// <value>
        /// CLU endpoint where application is hosted.
        /// </value>
        public string Endpoint { get; set; } = default!;

        /// <summary>
        /// Gets or sets CLU deployment name.
        /// </summary>
        /// <value>
        /// CLU deployment name.
        /// </value>
        public string DeploymentName { get; set; } = default!;
    }
}

File: packages/Recognizers/ConversationLanguageUnderstanding/js/src/cluApplication.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * Data describing a CLU application.
 */
export class CluApplication {
  /**
   * Initializes a new instance of the CluApplication class.
   * @param projectName CLU project name.
   * @param endpointKey CLU subscription or endpoint key.
   * @param endpoint CLU endpoint to use.
   * @param deploymentName CLU deployment name.
   */
  constructor(
    public projectName: string,
    public endpointKey: string,
    public endpoint: string,
    public deploymentName: string
  ) {
    if (!projectName?.trim()) {
      throw new Error(`CLU "projectName" parameter cannot be null or empty.`);
    }

    if (!this.isWellFormatedGUID(endpointKey)) {
      throw new Error(`"${endpointKey}" is not a valid CLU subscription key.`);
    }

    if (!endpoint?.trim()) {
      throw new Error(`CLU "endpoint" parameter cannot be null or empty.`);
    }

    if (!this.isWellFormedUriString(endpoint)) {
      throw new Error(`"${endpoint}" is not a valid CLU endpoint.`);
    }

    if (!deploymentName?.trim()) {
      throw new Error(
        `CLU "deploymentName" parameter cannot be null or empty.`
      );
    }
  }

  /**
   * Check if the provided value is a well formated GUID.
   * @param guid The GUID value.
   * @returns True if the GUID is well formated.
   */
  private isWellFormatedGUID(guid: string) {
    const pattern = /^(((?=.*}$){)|((?!.*}$)))((?!.*-.*)|(?=(.*[-].*){4}))[0-9a-fA-F]{8}[-]?([0-9a-fA-F]{4}[-]?){3}[0-9a-fA-F]{12}?[}]?$/m;
    return !!guid.match(pattern);
  }

  /**
   * Check if the provided value is a well formated URI.
   * @param uri The URI string value.
   * @returns True if the URI is well formated.
   */
  private isWellFormedUriString(uri: string): boolean {
    try {
      const uriResult = new URL(uri);
      return (
        (uriResult.toString() === uri || uriResult.toString() === `${uri}/`) &&
        (uriResult.protocol === 'https:' || uriResult.protocol === 'http:')
      );
    } catch {
      return false;
    }
  }
}
